kube-core/src/labels.rs

```diff
@@ -0,0 +1,649 @@
+#![allow(missing_docs)]
+use core::fmt;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::{LabelSelector, LabelSelectorRequirement};
+use serde::{Deserialize, Serialize};
+use std::{
+    cmp::PartialEq,
+    collections::{BTreeMap, BTreeSet},
+    fmt::Display,
+    iter::FromIterator,
+    option::IntoIter,
+};
+use thiserror::Error;
+
+use crate::ResourceExt;
+
+#[derive(Debug, Error)]
+#[error("failed to parse value as expression: {0}")]
+/// Indicates failure of conversion to Expression
+pub struct ParseExpressionError(pub String);
+
+// local type aliases
+type Expressions = Vec<Expression>;
+
+/// Extensions to [`ResourceExt`](crate::ResourceExt)
+/// Helper methods for resource selection based on provided Selector
+pub trait SelectorExt {
+    fn selector_map(&self) -> &BTreeMap<String, String>;
+
+    /// Perform a match on the resource using Matcher trait
+    ///
+    /// ```
+    /// use k8s_openapi::api::core::v1::Pod;
+    /// use k8s_openapi::apimachinery::pkg::apis::meta::v1::LabelSelector;
+    /// use kube_core::SelectorExt;
+    /// use kube_core::Expression;
+    /// let matches = Pod::default().selector_matches(&LabelSelector::default());
+    /// assert!(matches);
+    /// let matches = Pod::default().selector_matches(&Expression::Exists("foo".into()));
+    /// assert!(!matches);
+    /// ```
+    fn selector_matches(&self, selector: &impl Matcher) -> bool {
+        selector.matches(self.selector_map())
+    }
+}
+
+impl<R: ResourceExt> SelectorExt for R {
+    fn selector_map(&self) -> &BTreeMap<String, String> {
+        self.labels()
+    }
+}
+
+/// Matcher trait for implementing alternalive Selectors
+pub trait Matcher {
+    // Perform a match check on the resource labels
+    fn matches(&self, labels: &BTreeMap<String, String>) -> bool;
+}
```

Comment:
does this have to be public?

Reply:
I’m actually using this in https://github.com/crust-gather/crust-gather/pull/81/files#diff-870f123ca5b1ea5358ef5aa085d9b8a09b430f79459c3b1e320ed9312f3e191cR17 where the `Selector` wrapper implements the trait. I think this will break it, no?

Reply:
yep. let's discuss Matcher improvements separately since it'll be the one pub trait now. resolving.